AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxTest.java

```diff
@@ -48,141 +49,191 @@
+
+        assertEquals("4682943996362752##2", message2.getMessageId());
+        assertEquals("This is a second test message", message2.getTitle());
+        assertEquals("This is a second test message subtitle", message2.getSubtitle());
+        assertNull(message2.getExpirationTimestamp());
+        assertFalse(message2.isRead());
+      }
+    };
+
+    // Add callback to be able to validate.
+    Leanplum.getInbox().addChangedHandler(callback);
+
+    // Remove it afterwards so we don't get callbacks anymore.
+    Leanplum.getInbox().removeChangedHandler(callback);
+
+    // Validate message state.
+    List<LeanplumInboxMessage> messageList = Leanplum.getInbox().unreadMessages();
+
+    LeanplumInboxMessage message1 = messageList.get(0);
+    LeanplumInboxMessage message2 = messageList.get(1);
+
+    message1.read();
+    assertTrue(message1.isRead());
+    assertEquals(1, Leanplum.getInbox().unreadCount());
+
+    message2.read();
+    assertTrue(message2.isRead());
+    assertEquals(0, Leanplum.getInbox().unreadCount());
+
+    assertEquals(2, Leanplum.getInbox().count());
+
+    LeanplumInboxMessage messageById = Leanplum.getInbox().messageForId(message1.getMessageId());
+    assertNotNull(messageById);
+    assertEquals(message1, messageById);
+
+    Leanplum.getInbox().removeMessage(messageById.getMessageId());
+    assertEquals(1, Leanplum.getInbox().allMessages().size());
+
+    // Validate inbox synced callback with false when there is no internet .
+    InboxSyncedCallback inboxSyncedCallbackFalse = new InboxSyncedCallback() {
+      @Override
+      public void onForceContentUpdate(boolean success) {
+        assertFalse(success);
+      }
+    };
+
+    // Add synced callback  to be able to validate.
+    Leanplum.getInbox().addSyncedHandler(inboxSyncedCallbackFalse);
+
+    // Return false when checks for internet connection.
+    doReturn(false).when(Util.class, "isConnected");
+
+    // Download messages.
+    Leanplum.getInbox().downloadMessages();
+
+    //  Remove synced callback afterwards so we don't get synced callbacks anymore.
+    Leanplum.getInbox().removeSyncedHandler(inboxSyncedCallbackFalse);
+  }
+
+  @Test
+  public void testDisablePrefetching() {
+    leanplumInbox.disableImagePrefetching();
+    assertFalse(leanplumInbox.isInboxImagePrefetchingEnabled());
+  }
+
+  @Test
+  public void testInvalidMessageIdIsRejected() {
+    Date delivery = new Date(100);
+    Date expiration = new Date(200);
+    HashMap<String, Object> map = new HashMap<>();
+    map.put(Constants.Keys.MESSAGE_DATA, new HashMap<String, Object>());
+    map.put(Constants.Keys.DELIVERY_TIMESTAMP, delivery.getTime());
+    map.put(Constants.Keys.EXPIRATION_TIMESTAMP, expiration.getTime());
+    map.put(Constants.Keys.IS_READ, true);
+
+    LeanplumInboxMessage invalidMessage = LeanplumInboxMessage.createFromJsonMap("messageId", map);
+    assertNull(invalidMessage);
+  }
+
+  @Test
+  public void testRemoveMessageFromInbox() {
```

Comment:
why is this a separate method? you have it in testInbox (new lines 144-145)

Reply:
testInbox combines testing several behaviors and classes all into one giant test. This violates the Single Responsibility Principle, which all of our unit tests should respect. See this for reference: https://github.com/ghsukumar/SFDC_Best_Practices/wiki/F.I.R.S.T-Principles-of-Unit-Testing

Instead it is better to break our tests into smaller “unit tests” that cover only a single behavior. This way if we a commit breaks remove message from inbox functionality, we will fail this specific unit test and it will be easier to debug.